### PR TITLE
Enabling Normalization for DuckDB

### PR DESCRIPTION
**Pull request by sspaeti**
## What
Enabling Normalization that got disabled temporarily (see https://github.com/airbytehq/airbyte/pull/22528/files) and fix normalization integration tests. E.g. `setup_duck_db` missing.

Can be tested with:
```sh
NORMALIZATION_TEST_TARGET=duckdb  pytest integration_tests/test_normalization.py
```

## How
Adding `setup_duck_db` and fixing integration tests that failed.

## Recommended reading order
1. destination_definitions.yaml
2. dbt_integration_test.py

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Comment by sspaeti** [comment]
This PR will be revisited after the PR https://github.com/airbytehq/airbyte/pull/22381 is merged.

I added the missing `setup_duck_db` and the tests `./gradlew :airbyte-integrations:bases:base-normalization:airbyteDockerDuckDB ` are green:
```
BUILD SUCCESSFUL in 17s
67 actionable tasks: 14 executed, 53 up-to-date
S3 cache 663ms wasted on misses, reads: 1, elapsed: 663ms
```

But running `NORMALIZATION_TEST_TARGET=tidb  pytest integration_tests/test_normalization.py` directly, I see the error:
```
FAILED integration_tests/test_normalization.py::test_normalization[DestinationType.DUCKDB-test_nested_streams] - AssertionError: assert False
FAILED integration_tests/test_normalization.py::test_normalization[DestinationType.DUCKDB-test_simple_streams] - AssertionError: assert False
```

I'm unsure how to debug or fix these errors. Any help @edgao or @ryankfu would be appreciated when you have more air to breathe. What I noted thought, that I get the same errors for TIDB:

```
NORMALIZATION_TEST_TARGET=tidb  pytest integration_tests/test_normalization.py
...
FAILED integration_tests/test_normalization.py::test_normalization[DestinationType.TIDB-test_nested_streams] - AssertionError: assert False
FAILED integration_tests/test_normalization.py::test_normalization[DestinationType.TIDB-test_simple_streams] - AssertionError: assert False
```

So either it's a general error, or both have the same to fix. Any pointer or hint is greatly appreciated, as the [end-to-end test was successful](https://github.com/airbytehq/airbyte/pull/17494#issuecomment-1379086547).

File: airbyte-config/init/src/main/resources/seed/destination_definitions.yaml [hunk 1]
```diff
@@ -449,4 +449,9 @@
   dockerImageTag: 0.1.0
   documentationUrl: https://docs.airbyte.io/integrations/destinations/duckdb
   icon: duckdb.svg
+  normalizationConfig:
+    normalizationRepository: airbyte/normalization-duckdb
+    normalizationTag: 0.2.25
```
Comment (edgao) [inline airbyte-config/init/src/main/resources/seed/destination_definitions.yaml:454]:
```suggestion
    normalizationTag: 0.2.26
```

File: airbyte-integrations/bases/base-normalization/integration_tests/config.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"destination_path": "/local/test.duckdb"}
```
Comment (edgao) [inline airbyte-integrations/bases/base-normalization/integration_tests/config.json:1]: formatter seems to want a newline at the end of this file (probably easier to just `SUB_BUILD=CONNECTORS_BASE ./gradlew format`)

File: airbyte-integrations/bases/base-normalization/integration_tests/dbt_integration_test.py [hunk 3]
```diff
@@ -59,6 +59,8 @@ def setup_db(self, destinations_to_test: List[str]):
             self.setup_clickhouse_db()
         if DestinationType.TIDB.value in destinations_to_test:
             self.setup_tidb_db()
+        if DestinationType.DUCKDB.value in destinations_to_test:
```
Comment (edgao) [inline airbyte-integrations/bases/base-normalization/integration_tests/dbt_integration_test.py:62]: I'd be curious if you just don't have these two lines (+ delete the setup_duckdb_method) if tests still work. I think that would reduce test runtime, especially for local developers

**Review by edgao — COMMENTED** [review]
@ryankfu were there other items you wanted to bring up for duckdb? I think this pr looks reasonable, assuming tests pass (modulo comments) but I'm also pretty behind on this whole thing

**Comment by ryankfu** [comment]
@edgao Did not have any other remarks for getting DuckDB normalization back for the connector. Was mostly waiting on the fix for mitigating deeply nested objects and the freeze to be lifted before revisiting this PR but agree that this PR makes sense to me and reverts the filter logic you had in place

File: airbyte-integrations/bases/base-normalization/integration_tests/dbt_integration_test.py (hunk 3 above)
Comment (sspaeti) [reply airbyte-integrations/bases/base-normalization/integration_tests/dbt_integration_test.py]: I initially tried to remove, but I did it wrong, which resulted in this [error](https://github.com/airbytehq/airbyte/pull/22528#issue-1575062145). But I [removed](https://github.com/airbytehq/airbyte/pull/22566/commits/37c4e2cd6df5f3b1f0557809e9e4096ddf046023) it now as suggested, and the tests are all green still. Will resolve this convo.

**Comment by sspaeti** [comment]
I removed suggested integration tests and all is green. 

Testing locally - I can't activate normalization yet - how would I clear existing configs (in case they only get overwritten initially)? Last time I tried to delete local `actor_definition` table, does this still work:
image

## Minor Logo Question:
Plus a minor, the logo looks fine when building locally, but didn't when merged in master. Will try again after merge.
This is how it looks building locally:
image

and this from master:
image

**Comment by edgao** [comment]
hmmmmmmmmmmmmmm

things are pretty different after the monorepo shift and I'm still learning the new world myself

asking the conn ops folks here https://airbytehq-team.slack.com/archives/C03VDJ4FMJB/p1677027104919629

**Comment by edgao** [comment]
something about this is being spooky on macos. The destination connector is complaining that `/local/test.duckdb` doesn't exist - which is expected, I think? Since we're relying on it to _create_ that file

rerunning on CI in case this is specific to macos, but I do think we should make the tests pass locally as well.

also fyi - you can run the normalization testcases locally. E.g. `pytest -v 'integration_tests/test_normalization.py::test_normalization[DestinationType.DUCKDB-test_simple_streams]'`, or `./gradlew :airbyte-integrations:bases:base-normalization:integrationTest` to run the full suite

logs are pretty confusing to parse, but this message (from the destination connector, i.e. before even running normalization) has the problem:
```json
{"type": "TRACE", "trace": {"type": "ERROR", "emitted_at": 1677108815431.992, "error": {"message": "Something went wrong in the connector. See the logs for more details.", "internal_message": "IO Error: Cannot open file \"/local/test.duckdb\": No such file or directory", "stack_trace": "Traceback (most recent call last):\n  File \"/airbyte/integration_code/main.py\", line 11, in <module>\n    DestinationDuckdb().run(sys.argv[1:])\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/destinations/destination.py\", line 119, in run\n    for message in output_messages:\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/destinations/destination.py\", line 113, in run_cmd\n    yield from self._run_write(config=config, configured_catalog_path=parsed_args.catalog, input_stream=wrapped_stdin)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/destinations/destination.py\", line 49, in _run_write\n    yield from self.write(config=config, configured_catalog=catalog, input_messages=input_messages)\n  File \"/airbyte/integration_code/destination_duckdb/destination.py\", line 64, in write\n    con = duckdb.connect(database=path, read_only=False)\nduckdb.IOException: IO Error: Cannot open file \"/local/test.duckdb\": No such file or directory\n", "failure_type": "system_error"}}}
```

and you can pull out the stacktrace:
```
Traceback (most recent call last):
  File "/airbyte/integration_code/main.py", line 11, in <module>
    DestinationDuckdb().run(sys.argv[1:])
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/destinations/destination.py", line 119, in run
    for message in output_messages:
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/destinations/destination.py", line 113, in run_cmd
    yield from self._run_write(config=config, configured_catalog_path=parsed_args.catalog, input_stream=wrapped_stdin)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/destinations/destination.py", line 49, in _run_write
    yield from self.write(config=config, configured_catalog=catalog, input_messages=input_messages)
  File "/airbyte/integration_code/destination_duckdb/destination.py", line 64, in write
    con = duckdb.connect(database=path, read_only=False)
duckdb.IOException: IO Error: Cannot open file "/local/test.duckdb": No such file or directory
```

**Comment by edgao** [comment]
ah, I forgot that the setup_duckdb method is supposed to create `secrets/duckdb.json` >.> which is why CI failed

**Comment by edgao** [comment]
> if you run with pytest

pytest will still run the `destination-duckdb` container, it doesn't run the destination code directly. `./gradlew :airbyte-integrations:bases:base-normalization:airbyteDockerDuckDB` doesn't run the integration tests, it only runs the unit tests

I'll try and set this up locally tomorrow (mostly been dealing with oc stuff for the past few days...). If it works then we should make a separate pr with just the destination_definitions.yaml change (i.e. enable duckdb normalization in OSS) - everything else is just to keep the normalization build green, so we shouldn't merge it until it succeeds

**Comment by edgao** [comment]
hm, I finally got a sync to run (followed instructions [here](https://docs.airbyte.com/connector-development/testing-connectors/testing-a-local-catalog-in-development/) to get the deployment working) but it didn't seem to quite work as expected. The source emitted nonzero records:
```
2023-03-08 19:41:34 source > Read 1341 records from payment_intents stream
```
but the raw table was empty (which is weird? b/c this is just the connector itself, we haven't even run normalization yet)
```
D select * from _airbyte_raw_payment_intents;
┌────────────────┬─────────────────────┬───────────────┐
│ _airbyte_ab_id │ _airbyte_emitted_at │ _airbyte_data │
│    varchar     │        json         │     json      │
├──────────────────────────────────────────────────────┤
│                        0 rows                        │
└──────────────────────────────────────────────────────┘
```

and then normalization failed:
```
2023-03-08 19:43:36 normalization > Completed with 1 error and 0 warnings:
2023-03-08 19:43:36 normalization > Runtime Error in model payment_intents (models/generated/airbyte_tables/main/payment_intents.sql)
2023-03-08 19:43:36 normalization >   Binder Error: Catalog "main" does not exist!
```

this was using `destination-duckdb:0.1.0` + `normalization-duckdb:0.2.26`, i.e. what's available on the master branch.

full logs [bda075de_9bc7_4495_a31a_60dff5d8bcf7_logs_53_txt.txt](https://github.com/airbytehq/airbyte/files/10924541/bda075de_9bc7_4495_a31a_60dff5d8bcf7_logs_53_txt.txt)

also - it looks like duckdb has already released a new version, with a different storage format, do we need to release a new version of destination-duckdb?
```
$ ./duckdb /tmp/airbyte_local/test.duckdb
\Error: unable to open database "/tmp/airbyte_local/test.duckdb": IO Error: Trying to read a database file with version number 39, but we can only read version 43.
The database file was created with DuckDB version v0.6.0 or v0.6.1.
```

**Comment by sspaeti** [comment]
Hi @edgao 
Thanks so much for testing this.

Which source connector do you use? And how does your duckdb path look like? `/local/test.duckdb` or similar? And if so, was that file created in `/tmp/local_airbyte/test.duckdb`?

I normally use Faker which worked with and without normalization at some point.

Thanks for the [pointer](https://docs.airbyte.com/connector-development/testing-connectors/testing-a-local-catalog-in-development/), I tried to build this feature branch with these instructions. Unfortunately, they do not work for me (see Fig. 1).

## Versioning
Regarding new versions of DuckDB, that's a good question that I asked myself as well. DuckDB is updating quite often (ever[y 3-4 weeks](https://github.com/duckdb/duckdb/releases)). The best would be if we would somehow automatically build different docker images for different versions and this could be specified inside the settings of the destination (same as schema, path, etc.). Not sure if that is feasible. For now, we'd need manually 

Fig. 1:
image

**Comment by edgao** [comment]
interesting. I was using just `test.duckdb`, which passed connection tests (and generated the `/tmp/airbyte_local/test.duckdb` file correctly). Switching to `/local/test.duckdb` fixed it. That feels like a footgun - at minimum we should document it, and ideally this would cause the `check` method to fail.  (or `write` should be able to handle this case)

confirmed that the destination is able to write to the raw table, using `/local/test.duckdb`.

but now I'm running into errors with the storage version in normalization :/ it looks like the connector is writing a version 0.6.x duckdb file, but normalization wants to read 0.7.x?
```
2023-03-09 16:16:22 normalization > Encountered an error:
Runtime Error
  IO Error: Trying to read a database file with version number 39, but we can only read version 43.
  The database file was created with DuckDB version v0.6.0 or v0.6.1.
  
  The storage of DuckDB is not yet stable; newer versions of DuckDB cannot read old database files and vice versa.
  The storage will be stabilized when version 1.0 releases.
  
  For now, we recommend that you load the database file in a supported version of DuckDB, and use the EXPORT DATABASE command followed by IMPORT DATABASE on the current version of DuckDB.
  
  See the storage page for more information: https://duckdb.org/internals/storage
```

unclear why this is happening - I'm using our official images, not even locally-built versions:
```
2023-03-09 16:13:32 INFO i.a.c.i.LineGobbler(voidCall):149 - Checking if airbyte/destination-duckdb:0.1.0 exists...
<snip...>
2023-03-09 16:15:09 INFO i.a.w.t.s.NormalizationActivityImpl(lambda$normalize$3):166 - Using normalization: airbyte/normalization-duckdb:0.2.26
```

**Comment by edgao** [comment]
I _think_ it's expected that the gradle build will create new images? B/c it needs to pull in the new catalog json file? not super familiar with how this process works :/

**Comment by sspaeti** [comment]
That's a good catch. It handles when the path doesn't start with `/local` (see Fig. 1), but it doesn't when there is no path (I should add that yes).

## Version error

I can't test end to end, but I committed the versions to fix for `airbyte/destination-duckdb` and `airbyte/normalization-duckdb`. When building with Gradle locally, versions look good on my end (but images would need to be pushed again or built locally).

images built with Gradle:
```
./gradlew :airbyte-integrations:connectors:destination-duckdb:build
./gradlew :airbyte-integrations:bases:base-normalization:airbyteDockerDuckDB
docker image ls | grep duckdb
airbyte/normalization-duckdb         dev        c1299097b448   12 minutes ago   867MB
airbyte/destination-duckdb           dev        43cd23a0fa4e   6 hours ago      929MB
```

DuckDB version in destination image:
```sh
docker run -it --entrypoint /bin/bash airbyte/destination-duckdb:dev
root@27da9d164eda:/airbyte/integration_code# pip freeze | duckdb
bash: duckdb: command not found
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
BrokenPipeError: [Errno 32] Broken pipe
root@27da9d164eda:/airbyte/integration_code# pip freeze | grep duckdb
destination-duckdb @ file:///airbyte/integration_code
duckdb==0.7.1
```

And DuckDB and dbt-duckdb versions in normalization image:
```sh
docker run -it --entrypoint /bin/bash airbyte/normalization-duckdb:dev
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
root@d58b42ca7926:/airbyte# pip freeze | grep duckdb
dbt-duckdb==1.4.0
duckdb==0.7.1
```

Fig. 1:
image

**Comment by edgao** [comment]
hm, that commit is updating the normalization image? i.e. I'm guessing you meant to push a different commit for the connector? (and yeah, we'll need to `/publish` whichever one gets updated)

though... maybe it's more correct to pin both the connector and normalization to the same version? Because otherwise syncs could fail (if we publish one but not the other)

**Comment by edgao** [comment]
also, no idea what's up with you running platform :( just to doublecheck, here's pretty much my exact process:
1. `cd ~/code/airbyte`
2. `SUB_BUILD=ALL_CONNECTORS ./gradlew :airbyte-config:specs:generateOssConnectorCatalog`
3. `cp airbyte-config/init/src/main/resources/seed/oss_catalog.json ~/code/airbyte-platform-internal/oss/airbyte-config/init/src/main/resources/seed/local_dev_catalog.json`
    1. note the `airbyte-platform-internal/oss` - I think the docs page needs minor tweaking
4. `cd ~/code/airbyte-platform-internal/oss` (again, this is in the `oss` subdir)
5. `./gradlew build -x test`
6. `VERSION=dev docker compose up`

**Pull request closed without merging**